### PR TITLE
Fix #526 - Don't convert `|>` to a function call unless the RHS is a valid argument for a function call

### LINKED ISSUE #526: Parse Error when running JuliaFormatter at Yao master branch
running `format(".", BlueStyle())` gives the following error currently on Yao's master branch, any idea what happened?

---

<details>
<summary> Error message </summary>
julia> format(".", BlueStyle())
ERROR: Parsing error for formatted text:

using YaoBase, YaoArrayRegister, SimpleTraits

"""
    apply!(register, block)

Apply a block (of quantum circuit) to a quantum register.
"""
function apply!(r::AbstractRegister, b::AbstractBlock)
    _check_size(r, b)
    return _apply!(r, b)
end

function _apply!(r::AbstractRegister, b::AbstractBlock)
    return _apply_fallback!(r, b)
end

function _apply_fallback!(r::AbstractRegister, b::AbstractBlock)
    return throw(NotImplementedError(:_apply_fallback!, (r, b)))
end

function _apply_fallback!(r::ArrayReg{B,T}, b::AbstractBlock) where {B,T}
    r.state .= mat(T, b) * r.state
    return r
end

"""
    |>(register, blk)

Pipe operator for quantum circuits.

# Example

```julia
julia> ArrayReg(bit"0") |> X |> Y
```

!!! warning

    `|>` is equivalent to [`apply!`](@ref), which means it has side effects. You
    need to copy original register, if you do not want to change it in-place.
"""
Base.:)(()(r::AbstractRegister, blk::AbstractBlock) = apply!(r, blk)

function apply!(r::AbstractRegister, blk::Function)
    if applicable(blk, nactive(r))
        return apply!(r, blk(nactive(r)))
    else
        error(
            "input function is not applicable, it should take a integer as number of current active qubits.",
        )
    end
end

"""
    occupied_locs(x)

Return a tuple of occupied locations of `x`.
"""
occupied_locs(x::AbstractBlock) = (1:nqubits(x)...,)

"""
    subblocks(x)

Returns an iterator of the sub-blocks of a composite block. Default is empty.
"""
subblocks(x::AbstractBlock)

"""
    chsubblocks(composite_block, itr)

Change the sub-blocks of a [`CompositeBlock`](@ref) with given iterator `itr`.
"""
chsubblocks(x::AbstractBlock, itr)

"""
    applymatrix(g::AbstractBlock) -> Matrix

Transform the apply! function of specific block to dense matrix.
"""
function applymatrix(T, g::AbstractBlock)
    return linop2dense(T, r -> statevec(apply!(ArrayReg(r), g)), nqubits(g))
end
applymatrix(g::AbstractBlock) = applymatrix(ComplexF64, g)
# just use BlockMap maybe? No!

print_block(io::IO, blk::AbstractBlock) = print_block(io, MIME("text/plain"), blk)
print_block(blk::AbstractBlock) = print_block(stdout, blk)
print_block(io::IO, ::MIME"text/plain", blk::AbstractBlock) = summary(io, blk)

# return itself by default
Base.copy(x::AbstractBlock) = x

function cache_key end

"""
    mat([T=ComplexF64], blk)

Returns the matrix form of given block.
"""
mat(x::AbstractBlock) = mat(promote_type(ComplexF64, parameters_eltype(x)), x)

mat_matchreg(reg::AbstractRegister, x::AbstractBlock) = mat(x)
mat_matchreg(reg::ArrayReg{B,T}, x::AbstractBlock) where {B,T} = mat(T, x)

Base.Matrix{T}(x::AbstractBlock) where {T} = Matrix(mat(T, x))

# YaoBase interface
YaoBase.nqubits(::Type{<:AbstractBlock{N}}) where {N} = N
YaoBase.nqubits(x::AbstractBlock{N}) where {N} = nqubits(typeof(x))

# properties
for each_property in [:isunitary, :isreflexive, :ishermitian]
    @eval YaoBase.$each_property(x::AbstractBlock) = $each_property(mat(x))
    @eval function YaoBase.$each_property(::Type{T}) where {T<:AbstractBlock}
        return $each_property(mat(T))
    end
end

function iscommute_fallback(op1::AbstractBlock{N}, op2::AbstractBlock{N}) where {N}
    if length(intersect(occupied_locs(op1), occupied_locs(op2))) == 0
        return true
    else
        return iscommute(mat(op1), mat(op2))
    end
end

function YaoBase.iscommute(op1::AbstractBlock{N}, op2::AbstractBlock{N}) where {N}
    return iscommute_fallback(op1, op2)
end

# parameters
"""
    getiparams(block)

Returns the intrinsic parameters of node `block`, default is an empty tuple.
"""
getiparams(x::AbstractBlock) = ()

"""
    setiparams!([f], block, itr)
    setiparams!([f], block, params...)

Set the parameters of `block`.
When `f` is provided, set parameters of `block` to the value in `collection` mapped by `f`.
`iter` can be an iterator or a symbol, the symbol can be `:zero`, `:random`.
"""
function setiparams! end

"""
    setiparams([f], block, itr)
    setiparams([f], block, params...)

Set the parameters of `block`, the non-inplace version.
When `f` is provided, set parameters of `block` to the value in `collection` mapped by `f`.
`iter` can be an iterator or a symbol, the symbol can be `:zero`, `:random`.
"""
function setiparams end

for F in [:setiparams!, :setiparams]
    @eval begin
        function $F(x::AbstractBlock, args...)
            return if niparams(x) == length(args) == 0
                x
            else
                throw(NotImplementedError($(QuoteNode(F)), (x, args...)))
            end
        end

        $F(x::AbstractBlock, it::Union{Tuple,AbstractArray,Base.Generator}) = $F(x, it...)
        function $F(x::AbstractBlock, a::Number, xs::Number...)
            return error("setparams!(x, θ...) is not implemented")
        end
        $F(x::AbstractBlock, it::Symbol) = $F(x, render_params(x, it))

        function $F(f::Function, x::AbstractBlock, it)
            return $F(x, map(x -> f(x...), zip(getiparams(x), it)))
        end
        $F(f::Nothing, x::AbstractBlock, it) = $F(x, it)
        $F(f::Function, x::AbstractBlock, it::Symbol) = $F(f, x, render_params(x, it))
    end
end

"""
    parameters(block)

Returns all the parameters contained in block tree with given root `block`.
"""
parameters(x::AbstractBlock) = parameters!(parameters_eltype(x)[], x)

"""
    parameters!(out, block)

Append all the parameters contained in block tree with given root `block` to
`out`.
"""
function parameters!(out, x::AbstractBlock)
    append!(out, getiparams(x))
    for blk in subblocks(x)
        parameters!(out, blk)
    end
    return out
end

# = prewalk(blk->append!(out, getiparams(blk)), x)

niparams(x::AbstractBlock) = length(getiparams(x))

function nparameters(x::AbstractBlock)
    count = niparams(x)
    for each in subblocks(x)
        count += nparameters(each)
    end
    return count
end

"""
    iparams_eltype(block)

Return the element type of [`getiparams`](@ref).
"""
iparams_eltype(x::AbstractBlock) = eltype(getiparams(x))

"""
    parameters_eltype(x)

Return the element type of [`parameters`](@ref).
"""
function parameters_eltype(x::AbstractBlock)
    T = iparams_eltype(x)
    for each in subblocks(x)
        T = promote_type(T, parameters_eltype(each))
    end
    return T
end

mutable struct Dispatcher{VT}
    params::VT
    loc::Int
end

Dispatcher(params) = Dispatcher(params, 0)

function consume!(d::Dispatcher, n::Int)
    d.loc += n
    return d.params[(d.loc - n + 1):(d.loc)]
end

function consume!(d::Dispatcher{<:Symbol}, n::Int)
    d.loc += n
    return d.params
end

function consume!(d::Dispatcher{<:Number}, n::Int)
    if n == 0
        return ()
    elseif n == 1
        d.loc += n
        return d.params
    else
        error("do not have enough parameters to consume, expect 0, 1, got $n")
    end
end

function dispatch!(f::Union{Function,Nothing}, x::AbstractBlock, it::Dispatcher)
    setiparams!(f, x, consume!(it, niparams(x)))
    for each in subblocks(x)
        dispatch!(f, each, it)
    end
    return x
end

"""
    dispatch!(x::AbstractBlock, collection)

Dispatch parameters in collection to block tree `x`.

!!! note

    it will try to dispatch the parameters in collection first.
"""
function dispatch!(f::Union{Function,Nothing}, x::AbstractBlock, it)
    dp = Dispatcher(it)
    res = dispatch!(f, x, dp)
    @assert (it isa Symbol || length(it) == dp.loc) "expect $(dp.loc) parameters, got $(length(it))"
    return res
end

dispatch!(x::AbstractBlock, it) = dispatch!(nothing, x, it)

"""
    popdispatch!(f, block, list)

Pop the first [`nparameters`](@ref) parameters of list, map them with a function
`f`, then dispatch them to the block tree `block`. See also [`dispatch!`](@ref).
"""
function popdispatch!(f::Function, x::AbstractBlock, list::Vector)
    setiparams!(f, x, (popfirst!(list) for k in 1:niparams(x))...)
    for each in subblocks(x)
        popdispatch!(f, each, list)
    end
    return x
end

"""
    popdispatch!(block, list)

Pop the first [`nparameters`](@ref) parameters of list, then dispatch them to
the block tree `block`. See also [`dispatch!`](@ref).
"""
function popdispatch!(x::AbstractBlock, list::Vector)
    setiparams!(x, (popfirst!(list) for k in 1:niparams(x))...)
    for each in subblocks(x)
        popdispatch!(each, list)
    end
    return x
end

render_params(r::AbstractBlock, params) = params
render_params(r::AbstractBlock, params::Symbol) = render_params(r, Val(params))
render_params(r::AbstractBlock, ::Val{:random}) = (rand() for i in 1:niparams(r))
function render_params(r::AbstractBlock, ::Val{:zero})
    return (zero(iparams_eltype(r)) for i in 1:niparams(r))
end

"""
    HasParameters{X} <: SimpleTraits.Trait

Trait that block `X` has parameters.
"""
@traitdef HasParameters{X<:AbstractBlock}

@generated function SimpleTraits.trait(::Type{HasParameters{X}}) where {X}
    return hasmethod(parameters, Tuple{X}) ? :(HasParameters{X}) : :(Not{HasParameters{X}})
end

"""
    cache_type(::Type) -> DataType

Return the element type that a [`CacheFragment`](@ref)
will use.
"""
cache_type(::Type{<:AbstractBlock}) = Any

"""
    cache_key(block)

Returns the key that identify the matrix cache of this block. By default, we
use the returns of [`parameters`](@ref) as its key.
"""
cache_key(x::AbstractBlock)

function _check_size(r::AbstractRegister, pb::AbstractBlock{N}) where {N}
    return N == nactive(r) ||
        throw(QubitMismatchError("register size $(nactive(r)) mismatch with block size $N"))
end

"""
    parameters_range(block)

Return the range of real parameters present in `block`.

!!! note
    It may not be the case that `length(parameters_range(block)) == nparameters(block)`.

# Example

```jldoctest; setup=:(using YaoBlocks)
julia> parameters_range(RotationGate(X, 0.1))
ERROR: UndefVarError: parameters_range not defined
Stacktrace:
 [1] top-level scope
   @ none:1
```
"""
function parameters_range(block::AbstractBlock)
    T = parameters_eltype(block)
    out = Tuple{T,T}[]
    parameters_range!(out, block)
    return out
end

function parameters_range!(out::Vector{Tuple{T,T}}, block::AbstractBlock) where {T}
    for subblock in subblocks(block)
        parameters_range!(out, subblock)
    end
end

# non-inplace versions
"""
    apply(register, block)

The non-inplace version of applying a block (of quantum circuit) to a quantum register.
Check `apply!` for the faster inplace version.
"""
apply(r::AbstractRegister, b) = apply!(copy(r), b)

function generic_dispatch!(f::Union{Function,Nothing}, x::AbstractBlock, it::Dispatcher)
    x = setiparams(f, x, consume!(it, niparams(x)))
    return chsubblocks(
        x,
        map(subblocks(x)) do each
            generic_dispatch!(f, each, it)
        end,
    )
end

"""
    dispatch(x::AbstractBlock, collection)

Dispatch parameters in collection to block tree `x`, the generic non-inplace version.

!!! note

    it will try to dispatch the parameters in collection first.
"""
function dispatch(f::Union{Function,Nothing}, x::AbstractBlock, it)
    dp = Dispatcher(it)
    res = generic_dispatch!(f, x, dp)
    @assert (it isa Symbol || length(it) == dp.loc) "expect $(dp.loc) parameters, got $(length(it))"
    return res
end

dispatch(x::AbstractBlock, it) = dispatch(nothing, x, it)

 Error occured on line 423, offset 56.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] format_text(cst::CSTParser.EXPR, style::BlueStyle, s::JuliaFormatter.State)
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:581
  [3] format_text(text::String, style::BlueStyle, opts::JuliaFormatter.Options)
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:529
  [4] #format_text#204
    @ ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:521 [inlined]
  [5] format_text
    @ ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:519 [inlined]
  [6] #format_text#203
    @ ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:515 [inlined]
  [7] format_file(filename::String; overwrite::Bool, verbose::Bool, format_markdown::Bool, format_options::Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}})
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:645
  [8] (::JuliaFormatter.var"#209#212"{Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}}})(formatted_file::Bool, file::String)
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:729
  [9] BottomRF
    @ ./reduce.jl:81 [inlined]
 [10] _foldl_impl(op::Base.BottomRF{JuliaFormatter.var"#209#212"{Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}}}}, init::Bool, itr::Vector{String})
    @ Base ./reduce.jl:62
 [11] foldl_impl
    @ ./reduce.jl:48 [inlined]
 [12] mapfoldl_impl
    @ ./reduce.jl:44 [inlined]
 [13] _mapreduce_dim
    @ ./reducedim.jl:327 [inlined]
 [14] #mapreduce#725
    @ ./reducedim.jl:322 [inlined]
 [15] #reduce#727
    @ ./reducedim.jl:371 [inlined]
 [16] (::JuliaFormatter.var"#208#211"{Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}}})(formatted_path::Bool, dir_branch::Tuple{String, Vector{String}, Vector{String}})
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:717
 [17] BottomRF
    @ ./reduce.jl:81 [inlined]
 [18] _foldl_impl(op::Base.BottomRF{JuliaFormatter.var"#208#211"{Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}}}}, init::Bool, itr::Channel{Tuple{String, Vector{String}, Vector{String}}})
    @ Base ./reduce.jl:62
 [19] foldl_impl(op::Base.BottomRF{JuliaFormatter.var"#208#211"{Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}}}}, nt::Bool, itr::Channel{Tuple{String, Vector{String}, Vector{String}}})
    @ Base ./reduce.jl:48
 [20] mapfoldl_impl
    @ ./reduce.jl:44 [inlined]
 [21] #mapfoldl#244
    @ ./reduce.jl:162 [inlined]
 [22] #mapreduce#248
    @ ./reduce.jl:289 [inlined]
 [23] #reduce#250
    @ ./reduce.jl:458 [inlined]
 [24] format(paths::Tuple{String}; options::Base.Pairs{Symbol, BlueStyle, Tuple{Symbol}, NamedTuple{(:style,), Tuple{BlueStyle}}})
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:714
 [25] #format#213
    @ ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:739 [inlined]
 [26] #format#214
    @ ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:744 [inlined]
 [27] format(path::String, style::BlueStyle)
    @ JuliaFormatter ~/.julia/packages/JuliaFormatter/mSi5J/src/JuliaFormatter.jl:744
 [28] top-level scope
    @ REPL[3]:1
</details>

### PR DESCRIPTION
**Pull request by domluna**

### CODE REVIEW

File: test/issues.jl [hunk 1]
```diff
@@ -1240,10 +1240,15 @@
...
+        @test fmt(str_, pipe_to_function_call = true) == str
 
         str_ = "output = input .|> f.g.h"
         str = "output = f.g.h.(input)"
-        fmt(str_, pipe_to_function_call = true) == str
+        @test fmt(str_, pipe_to_function_call = true) == str
+    end
+
+    @testset "526" begin
+        str = "Base.:(|>)(r::AbstractRegister, blk::AbstractBlock) = apply!(r, blk)"
```
Comment (domluna) [inline test/issues.jl:1251]: corner case where the RHS is punctuation
